pyscraper/sp_2024/download.py

```diff
@@ -0,0 +1,225 @@
+"""
+Module for downloading the debates from the Scottish Parliament website.
+
+This uses the search page to get a list of committees that have met on a given date or range.
+Individual agenda items are then queried and dumped in an XML file in the cache directory.
+One file per comitee per day.
+
+"""
+
+from __future__ import annotations
+
+from itertools import groupby
+from pathlib import Path
+from typing import Iterator, NamedTuple
+from urllib.parse import parse_qs, urlparse
+
+import requests
+from bs4 import BeautifulSoup
+from lxml import etree
+
+user_agent = (
+    "Mozilla/5.0 (Ubuntu; X11; Linux i686; rv:9.0.1) Gecko/20100101 Firefox/9.0.1"
+)
+
+scot_prefix = "https://www.parliament.scot"
+search_url = "/chamber-and-committees/official-report/search-what-was-said-in-parliament?msp=&committeeSelect=&qry=&dateSelect=custom&dtDateFrom={iso_date}&dtDateTo={iso_date}&showPlenary=true&ShowDebates=true&ShowFMQs=true&ShowGeneralQuestions=true&ShowPortfolioQuestions=true&ShowSPCBQuestions=true&ShowTopicalQuestions=true&ShowUrgentQuestions=true&showCommittee=true&page=1"
+item_url = "/chamber-and-committees/official-report/search-what-was-said-in-parliament/{slug}?meeting={id}&iob={iob}"
+major_heading_url = "/chamber-and-committees/official-report/search-what-was-said-in-parliament/{slug}?meeting={id}"
+
+
+def get_meeting_urls(date: str, page: int = 1):
+    """
+    Query the Scottish Parliament search page for a given date to get all links for agenda meetings
+    """
+    date_url = scot_prefix + search_url.format(iso_date=date, page=page)
+    response = requests.get(date_url, headers={"User-Agent": user_agent})
+
+    # extract all urls that contain /search-what-was-said-in-parliament/
+    soup = BeautifulSoup(response.text, "html.parser")
+    meeting_urls = [
+        a["href"]
+        for a in soup.find_all("a", href=True)
+        if "/search-what-was-said-in-parliament/" in a["href"]
+    ]
+
+    # section headings
+    heading_urls = [url for url in meeting_urls if "&iob=" not in url]
+    heading_count = len(set(heading_urls))
+
+    # discard any without a iob parameter
+    meeting_urls = [url for url in meeting_urls if "&iob=" in url]
+
+    # remove duplicates
+    meeting_urls = sorted(list(set(meeting_urls)))
+    return heading_count, meeting_urls
+
+
+def get_debate_groupings(date: str) -> list[DebateGrouping]:
```

Comment:
Hopefully changed in 2940dbfa

Reply:
Yes - was just this way because I wrote the one day approach first. Possibly a knock-on issue in where the stored file gets it's current date? Although it's in the file name and so can be got from other places.

Reply:
I do add a `date = '-'.join(reversed(g.split('-')[-3:]))` in the groupby for this, which was the only place it was needed I think, dunno if missed somewhere.